### PR TITLE
Fixed inverse kinematics

### LINKED ISSUE #205: Rename motor position optimization to aim point optimization
We could consider renaming the motor position optimization to aim point optimization.

(surface reconstruction is also not named control point reconstruction and kinematics reconstruction is also not named actuator parameter reconstruction.)

### LINKED ISSUE #213: Combine all data files in the tests dir
Currently there are many separate files storing expected data for the unit tests. Most files contain a single tensor, we could store them all in one file, organizing them in a dict for example.

### PR DESCRIPTION
**Pull request by MarleneBusch**
# Description

The main feature of this PR is the fixed inverse kinematics.

- add second kinematics reconstruction method 
- add train test dataset splits for the kinematics reconstructor and surface reconstructor
- rename the motor position optimizer to aim point optimizer
- for the kinematics and surface reconstruction pytests, i have removed test heliostat AA31 and instead added one more sample to test heliostat AA39, to keep the compute load similar but increase the information gain in the reconstruction with very few samples.

Fixes #205 
Fixes #213 (there is now only one .pt file storing all tensors used in the pytests in a dict, this reduces 35 files into 1 file, it is slightly less userfriendly maybe, so if this change is not wanted I can revert it.)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: artist/field/kinematics_rigid_body.py [hunk 1]
```diff
@@ -43,11 +49,17 @@ class RigidBody(Kinematics):
...
-        Shape is ``[4]``.
     actuators : Actuators
         The actuators used in the kinematics.
+    kinematics_standard_orientation : torch.Tensor
+        Standard orientation of the kinematics system: south (0, -1, 0, 0) in homogeneous ENU.
+        Shape is ``[4]``.
+    initial_orientation_offsets : torch.Tensor
+        Rotation matrix to account for the initial orientation offset between surface mesh and kinematics system.
+        Shape is ``[1, 4, 4]``.
+    homogeneous_origin
```
Comment (mcw92) [inline artist/field/kinematics_rigid_body.py]:
```suggestion
    homogeneous_origin : torch.Tensor
```
I assume this is also a tensor?

File: artist/field/kinematics_rigid_body.py [hunk 2]
```diff
@@ -291,17 +323,28 @@ def _compute_orientations_from_motor_positions(
...
+        device: torch.device | None = None,
     ) -> torch.Tensor:
         """
-        Apply the initial orientation offsets to the given orientation matrices.
+        Compute the motor positions from a normal vector.
+
+        This is the inverse kinematics. First the joint angles are computed from the desired normal vector.
+        Then the motor positions are computed from the joint angles.
+        The inverse kinematics produces two solutions, the valid solution is chosen according to resulting
+        motor positions which must lie within the minimum and maximum allowed motor positions defined within
```
Comment (mcw92) [inline artist/field/kinematics_rigid_body.py]:
```suggestion
        motor positions which must lie within the minimum and maximum allowed motor positions defined in
```

File: artist/field/kinematics_rigid_body.py [hunk 3]
```diff
@@ -310,35 +353,193 @@ def _apply_initial_orientation_offsets(
...
+        # The forward kinematics model:
+        # n = R_n(τ1n) * R_u(τ1u) * R_e(θ1) * R_e(τ2e) * R_n(τ2n) * R_u(θ2) * ref
+        # where
+        # n                     : normal
+        # R                     : rotation matrices
+        # τ1n, τ1u, τ2e, τ2n    : rotation_deviations for both axes
+        # ref                   : reference direction of the kinematics system (0,-1,0) ENU
+        # θ1, θ2                : first_joint_angle, second_joint_angle
+        #
+        # This is inverted in the following.
```
Comment (mcw92) [inline artist/field/kinematics_rigid_body.py]:
```suggestion
        # This is inverted in the following, where
        # θ1 and θ2 are the two unknowns of this inverse problem we want to find out.
```

File: artist/field/kinematics_rigid_body.py [hunk 4]
```diff
@@ -310,35 +353,193 @@ def _apply_initial_orientation_offsets(
...
+
+        # Wrap both solutions into [-pi, pi].
+        second_joint_angle_1 = torch.atan2(
+            torch.sin(second_joint_angle_1), torch.cos(second_joint_angle_1)
+        )
+        second_joint_angle_2 = torch.atan2(
+            torch.sin(second_joint_angle_2), torch.cos(second_joint_angle_2)
+        )
+
+        # As the second joint angle is now known, the first joint angle can be computed using:
```
Comment (mcw92) [inline artist/field/kinematics_rigid_body.py]:
```suggestion
        # As the second joint angle is now known, the first joint angle can be computed using
```

File: artist/field/kinematics_rigid_body.py [hunk 5]
```diff
@@ -310,35 +353,193 @@ def _apply_initial_orientation_offsets(
...
+            first_rotation_axis_deviations.transpose(-1, -2) @ normals[:, :, None]
+        )[:, :, indices.squeeze_index]
+
+        # Reduced kinematics after removing first joint rotation deviations:
+        # n' = R_e(θ1) * v(θ2),      with v(θ2) := F2 R_u(θ2) * ref
+        # where: F2 = R_e(τ2e) * R_n(τ2n).
+        # As the first joint rotates around the e-axis, only the n- and u-components change and
+        # n'_e = v_e(θ2).
+        # Substituting v_e(θ2) back in and considering that the second joint rotates around the u-axis:
+        # n'_e = v_e(θ2) = F2_00 * cos(θ2) - F2_01 * sin(θ2) + F2_02 * 0
```
Comment (mcw92) [inline artist/field/kinematics_rigid_body.py]: I am a bit confused by this comment. In Max' document, this reads a bit different (and I think this is correct, at least in my understanding):

```suggestion
        # n'_e = v_e(θ2) = F2_00 * sin(θ2) - F2_01 * cos(θ2) + F2_02 * 0 = A * sin(θ2) + B * cos(θ2)
        # with A := F2_00 and B := - F2_01
```

File: artist/field/kinematics_rigid_body.py [hunk 6]
```diff
@@ -310,35 +353,193 @@ def _apply_initial_orientation_offsets(
...
+        # In the above equation only θ2 is unknown. It is solved by reducing the expression
+        # to a phase-shifted sinusoid: n'_e = a * sin(θ2 + φ) which results in two possible solutions.
```
Comment (mcw92) [inline artist/field/kinematics_rigid_body.py]:
```suggestion
        # In the above equation only θ2 is unknown. It is solved by reducing the expression
        # to a phase-shifted sinusoid via n'_e = a * sin(θ2 + φ) with a = sqrt(A^2 + B^2) and φ = atan2(B, A),
        # which results in two possible solutions:
```

File: artist/field/kinematics_rigid_body.py [hunk 7]
```diff
@@ -352,14 +553,14 @@ def incident_ray_directions_to_orientations(
...
-        max_num_iterations : int
-            Maximum number of iterations (default is 2).
-        min_eps : float
-            Convergence criterion (default is 0.0001).
         device : torch.device | None
             The device on which to perform computations or load tensors and models (default is None).
             If None, ``ARTIST`` will automatically select the most appropriate
             device (CUDA or CPU) based on availability and OS.
+        max_num_iterations : int
+            Maximum number of iterations (default is 2).
```
Comment (mcw92) [inline artist/field/kinematics_rigid_body.py]:
```suggestion
            Maximum number of iterations (default is 4).
```

File: artist/optim/kinematics_reconstructor.py [hunk 8]
```diff
@@ -154,25 +182,245 @@ def reconstruct_kinematics(
...
+                for spine in axes[sample_offset, :].flat:
+                    for spines in spine.spines.values():
+                        spines.set_edgecolor(border_color)
+                        spines.set_linewidth(4)
+
+            plt.tight_layout()
+            plt.savefig(f"heliostat_{heliostat_index}_{plot_name}")
+            plt.close(fig)
+
+    def _reconstruct_kinematics_parameters_wih_alignment(
```
Comment (mcw92) [inline artist/optim/kinematics_reconstructor.py]:
```suggestion
    def _reconstruct_kinematics_parameters_with_alignment(
```
Should probably rename / refactor properly in IDE.

File: artist/optim/kinematics_reconstructor.py [hunk 9]
```diff
@@ -154,25 +182,245 @@ def reconstruct_kinematics(
...
+            torch.mean(test_loss_kl_div).item(),
+        )
+
+        return flux_prediction, {
+            "pixel_loss": test_loss_pixel,
+            "kl_div": test_loss_kl_div,
+            "focal_spot_loss": test_loss_focal_spot,
+        }
+
+    def _plot_fluxes(
```
Comment (mcw92) [inline artist/optim/kinematics_reconstructor.py]: Not quite sure whether we should have plotting functionality in the actual package. What exactly do you use that for?

File: artist/optim/kinematics_reconstructor.py [hunk 10]
```diff
@@ -246,68 +493,307 @@ def _reconstruct_kinematics_parameters_with_raytracing(
...
+
+                    orientations = (
+                        heliostat_group.kinematics.motor_positions_to_orientations(
+                            motor_positions=data_split.motor_positions_train,
+                            device=device,
+                        )
+                    )
+
+                    normals_predicted = orientations @ torch.tensor(
+                        [0.0, 0.0, 1.0, 0.0], device=device
```
Comment (mcw92) [inline artist/optim/kinematics_reconstructor.py:478]: Should we also introduce a variable for the up reference [0.0, 0.0, 1.0, 0.0] (as in the inverse kinematics function for the south ref.)?

File: artist/optim/kinematics_reconstructor.py (hunk 8 above)
Comment (mcw92) [inline artist/optim/kinematics_reconstructor.py]: This method is quite long and does a lot of different things (data parsing, training loop, scheduling, validation, plotting, logging, DDP bookkeeping) in a single function. Can we do a bit of refactoring by extracting some pieces into helpers, e.g. `_prepare_group_calibration_data(...)`, `_compute_measured_normals(...)`, `_run_alignment_optimization_for_group(...)`, or `_update_final_loss_per_heliostat(...)`?

That would make the main method a clearer “orchestration shell” and potentially simplify testing for each stage.

In addition, I would clarify the normal conventions and geometry relationship used (similar as done for the inverse kinematics as mentioned again below).

File: artist/optim/kinematics_reconstructor.py [hunk 11]
```diff
@@ -246,68 +493,307 @@ def _reconstruct_kinematics_parameters_with_raytracing(
...
+                log.info(f"Rank: {rank}, Kinematics reconstructed.")
+
+        for heliostat_group in self.scenario.heliostat_field.heliostat_groups:
+            heliostat_group.kinematics.rotation_deviation_parameters = (
+                heliostat_group.kinematics.rotation_deviation_parameters.detach()
+            )
+
+        return final_loss_per_heliostat.detach().cpu(), [loss_history]
+
+    def _reconstruct_kinematics_parameters_with_raytracing(
```
Comment (mcw92) [inline artist/optim/kinematics_reconstructor.py:561]: This method might also benefit from extracting some parts into helpers for better readability.

File: artist/optim/loss.py [hunk 12]
```diff
@@ -157,7 +157,7 @@ def __call__(
         r"""
         Compute the focal spot loss.
 
-        First the focal spots of the prediction are computed, then the loss is computed and reduced
+        First the focal spots of the prediction and ground truth flux maps are computed, then the loss is computed and reduced
```
Comment (mcw92) [inline artist/optim/loss.py]:
```suggestion
        First the focal spots of the prediction and ground-truth flux maps are computed, then the loss is computed and reduced
```

File: artist/optim/loss.py [hunk 13]
```diff
@@ -436,26 +504,31 @@ def __call__(
...
+    Calculate the loss per heliostat from a loss per sample.
+
+    The reduction operation is chosen via the `reduction` parameter.
 
     Parameters
     ----------
     loss_per_sample : torch.Tensor
         Loss per sample.
-        Shape is ``[number_of_samples]``.
+        Tensor of shape [number_of_samples].
```
Comment (mcw92) [inline artist/optim/loss.py]:
```suggestion
        Shape is ``[number_of_samples]``.
```

File: artist/optim/loss.py [hunk 14]
```diff
@@ -436,26 +504,31 @@ def __call__(
...
         Number of samples per heliostat.
+    reduction : Callable[..., Any]
+        Reduction function applied across the sample dimension for each heliostat.
 
     Returns
     -------
     torch.Tensor
         Loss per heliostat.
-        Shape is ``[number_of_heliostats]``.
+        Tensor of shape [number_of_heliostats].
```
Comment (mcw92) [inline artist/optim/loss.py]:
```suggestion
        Shape is ``[number_of_heliostats]``.
```

File: artist/optim/surface_reconstructor.py [hunk 15]
```diff
@@ -139,12 +150,237 @@ def __init__(
...
+
+    def _validate(
+        self,
+        heliostat_group: HeliostatGroup,
+        data_split: training.TrainTestSplit,
+        evaluation_points: torch.Tensor,
+        device: torch.device | None = None,
+    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
+        """
+        Validate the surface reconstruction for a specified heliostat group on the test data.
```
Comment (mcw92) [inline artist/optim/surface_reconstructor.py:166]:
```suggestion
        Validate the surface reconstruction for a specified heliostat group on the test data.
        
        This validates a surface reconstruction (NURBS-based mirror surface shape + alignment) for a given heliostat group on the test set:

        - It builds NURBS surfaces from control points.
        - Samples surface points and normals.
        - Aligns those surfaces to incident rays and aim points.
        - Runs ray tracing to predict flux on the receiver.
        - Compares predicted vs measured flux for the test samples and computes several test losses.

        It returns the predicted flux images (for the local rank’s test samples) and a dict with per‑heliostat test losses (pixel loss and KL divergence).
```

File: tests/optim/test_loss_functions.py [hunk 16]
```diff
@@ -352,17 +325,15 @@ def test_pixel_loss(
                 ground_truth=ground_truth.to(device),
             )
         assert (
-            "The vector loss expects ['reduction_dimensions', 'device', 'target_area_indices'] as keyword arguments. Please add reduction_dimensions as keyword argument. Please add device as keyword argument. Please add target_area_indices as keyword argument."
+            "The vector loss expects ['reduction_dimensions'] as keyword arguments. Please add reduction_dimensions as keyword argument."
```
Comment (mcw92) [inline tests/optim/test_loss_functions.py]:
```suggestion
            "The vector loss expects ['reduction_dimensions'] as keyword arguments. Please add 'reduction_dimensions' as keyword argument."
```

File: tests/optim/test_loss_functions.py [hunk 17]
```diff
@@ -536,3 +507,62 @@ def test_angle_loss(
...
+    ----------
+    prediction : torch.Tensor
+        The predicted values.
+        Tensor of variable shape.
+    ground_truth : torch.Tensor
+        The ground truth.
+        Tensor of variable shape.
+    expected : torch.Tensor
+        The expected loss.
+        Tensor of shape [number_of_samples].
```
Comment (mcw92) [inline tests/optim/test_loss_functions.py]:
```suggestion
    Parameters
    ----------
    prediction : torch.Tensor
        The predicted values.
        Tensor of variable shape.
    ground_truth : torch.Tensor
        The ground truth.
        Tensor of variable shape.
    expected : torch.Tensor
        The expected loss.
        Tensor of shape ``[number_of_samples]``.
```

**Review by mcw92 — CHANGES_REQUESTED** [review]
LGTM, I think we are now very close to release-ready. A couple of things to be addressed before we do the merge:

- I checked the inverse kinematics and all related changes carefully. The implemented functionality matches the formulae provided by @Filos1992 and makes sense to me.

- I made some small changes for obvious things, like minor modifications to docstrings or comments etc. The commit suggestions I was not sure about I left unchanged to be addressed once @MarleneBusch is back from vacation. 

- Some methods (I made specific comments there) would really benefit from further refactoring into helper functions to improve clarity and readability of the code.

- I am very unsure about whether we should have plotting functionality implemented at such a deep level inside the package as it is the case for the `KinematicsReconstructor` class. Can we move these "analysis" methods somewhere else?

- I think we should really move all scripts related to the full-field optimization to another repo inside the ARTIST Association that just imports `artist` as a package. We should have done the same for the PAINT plots; however, we cannot move these scripts anymore as they are explicitly linked in the Nature Energy paper.

Note that I did not check the plotting and example scripts properly.

File: artist/optim/kinematics_reconstructor.py (hunk 8 above)
Comment (MarleneBusch) [reply artist/optim/kinematics_reconstructor.py]: done

File: artist/optim/kinematics_reconstructor.py (hunk 9 above)
Comment (MarleneBusch) [reply artist/optim/kinematics_reconstructor.py]: the plotting has been removed

File: artist/optim/kinematics_reconstructor.py (hunk 10 above)
Comment (MarleneBusch) [reply artist/optim/kinematics_reconstructor.py:478]: Refactoring this is a good idea and will happen in another PR.

File: artist/optim/kinematics_reconstructor.py (hunk 8 above)
Comment (MarleneBusch) [reply artist/optim/kinematics_reconstructor.py]: Refactoring this is a good idea and will happen in another PR.

File: artist/optim/kinematics_reconstructor.py (hunk 11 above)
Comment (MarleneBusch) [reply artist/optim/kinematics_reconstructor.py:561]: Refactoring this is a good idea and will happen in another PR.

**Review by mcw92 — APPROVED** [review]
